packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -578,6 +648,97 @@ export class TokenDetectionController extends StaticIntervalPollingController<To
     return slicesOfTokensToDetect;
   }
 
+  /**
+   * This adds detected tokens from the Accounts API, avoiding the multi-call RPC calls for balances
+   * @param options - method arguments
+   * @param options.selectedAddress - address to check against
+   * @param options.chainId - chainId to check tokens for
+   * @param options.tokenCandidateSlices - these are tokens we know a user does not have (by checking the tokens controller).
+   * We will use these these token candidates to determine if a token found from the API is valid to be added on the users wallet.
+   * It will also prevent us to adding tokens a user already has
+   * @returns a success or failed object
+   */
+  async #addDetectedTokensViaAPI({
+    selectedAddress,
+    chainId,
+    tokenCandidateSlices,
+  }: {
+    selectedAddress: string;
+    chainId: Hex;
+    tokenCandidateSlices: string[][];
+  }) {
+    return await safelyExecute(async () => {
+      const tokenBalances = await this.#accountsAPI
+        .getMultiChainBalances(selectedAddress, chainId)
+        .catch(() => null);
+
+      if (!tokenBalances || tokenBalances.length === 0) {
+        return { result: 'failed' } as const;
+      }
+
+      const tokensWithBalance: Token[] = [];
+      const eventTokensDetails: string[] = [];
+
+      const tokenCandidateSet = new Set<string>(tokenCandidateSlices.flat());
+
+      tokenBalances.forEach((token) => {
+        const tokenAddress = token.address;
+
+        // Make sure that the token to add is in our candidate list
+        // Ensures we don't add tokens we already own
+        if (!tokenCandidateSet.has(token.address)) {
+          return;
+        }
+
+        // We need specific data from tokenList to correctly create a token
```

Comment:
Are the fields we get from the token list the icon url and aggregators?

We'll eventually stop storing icon urls in state since its always calculable as `"https://static.cx.metamask.io/api/v1/tokenIcons/${chainId}/${address}.png"`.

Aggregators are shown in the UI but we could just query them from the big token list rather than duplicating on each imported token.  So I think we keep this for now but could remove the dependency later.

Reply:
Aha saw a relavant comment you've made before, great callout.
https://github.com/MetaMask/mobile-planning/issues/1935#issuecomment-2384061477